### PR TITLE
Feature add spot mini robot

### PR DESCRIPTION
**Pull request by DavidMansolino**
**Description**
This PR replaces #1461, by migrating the source branch on this repo instead of a fork.

**Remaining Tasks**
- [x] Update with master.
- [x] _Controller_:
  - [x]  Finish the `go_forward()` function
  - [x]  Clang-format.
- [x] _Documentation:_ 
  - [x] Generate icon and model
  - [x] thumbnails images
  - [x] WEB component
- [x] _New Appearances:_ 
  - [x] https://github.com/cyberbotics/webots/tree/feature-add-SpotMini-robot/projects/robots/boston_dynamics/spot/protos/textures/holes_perforated_steel duplicated with https://github.com/cyberbotics/webots/blob/feature-add-SpotMini-robot/projects/samples/environments/factory/worlds/textures/metal_grid_base_color.png
- [x] _Tests:_ Fix Travis errors
- [x] _Improvement:_ Find a solution to have a real "piston" behaviour in the elbow or remove the `DEF TENDON Solid` node.

### CODE REVIEW

**Comment by DavidMansolino** [comment]
I have made some cleanups to this PR, the two main points remaining to fix are:
- [ ] The walking algorithm is not yet working.
- [ ] The robot is not well centred (centre of the robot completely outside of the robot).

**Comment by DavidMansolino** [comment]
My latest commit fixes most of the solid and joint origin but unfortunately it breaks completely the controller, because the joints have moved a lot (they were not inside the robot at all).

**Comment by DavidMansolino** [comment]
I have been able to simplify the controller and make it way more robust but there is still something not realistic at all with the behaviour of this robot, its motions does not make sense at all (it looks like the robot is in water or something similar).
For now, I suspect two possible causes:
  - The bounding objects may be too complex.
  - The mass distribution may be wrong.

**Comment by omichel** [comment]
Can you post a short movie about it?

**Comment by DavidMansolino** [comment]
Here is a video showing the problem:
https://www.youtube.com/watch?v=FdhAOyV8VpY&feature=youtu.be
Their is no spring or damping effect, front and rear legs are doing the exact same motion.

**Comment by DavidMansolino** [comment]
I checked the masses distribution centre of masses and inertia and they are correct.
I also tried to simplify a lot the key bounding objects.
But it doesn't impact the behaviour.

I am now suspecting the legs definition, it seems in the leg too there is coordinate issue (the origin of the leg is completely outside of the leg itself (by more than 1.5 meters), even if the bounding objects, joints and physics are defined accordingly I suspect it may cause issues to ODE.

**Comment by omichel** [comment]
To me, it looks like the mass is too lightweight or the strength of the motors is too high...

**Comment by DavidMansolino** [comment]
The mass is the one of the doc (with is more than 25kg so not so lightweight), but good points about the motors, I will check this.

**Comment by DavidMansolino** [comment]
About the max torque of the motors, it is set to 50Nm per motor, which considering 4 legs and a total mass of 25-30kg seems to make sense.

**Comment by DavidMansolino** [comment]
Correcting the legs is a real nightmare, it is extremely time consuming and error prone 😞

**Comment by DavidMansolino** [comment]
I have fixed all the legs (and sub-parts) origin and it is now working like a charm (realistic motion and much more stable):
https://youtu.be/XI4R89o7yRw

**Comment by DavidMansolino** [comment]
✔️: travis_release (https://travis-ci.com/github/cyberbotics/webots/builds/161386660)

**Comment by omichel** [comment]
:eyes:

File: docs/guide/spot.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+Derived from [Robot](../reference/robot.md).
+
+```
+Spot {
+  SFVec3f     translation      0 -1.028 0.46
+  SFRotation  rotation         0 1 0 0
+  SFString    name             "Spot"
+  SFString    model            "Boston Dynamics - Spot"
+  SFString    controller       "moving_demo"
```
Comment (omichel) [inline docs/guide/spot.md:24]:
```suggestion
  SFString    controller       "spot_moving_demo"
```

File: projects/robots/boston_dynamics/spot/controllers/spot_moving_demo/spot_moving_demo.c [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+#define NUMBER_OF_JOINTS 12
+#define NUMBER_OF_CAMERAS 5
+
+// Initialize the robot's information
+static WbDeviceTag motors[NUMBER_OF_JOINTS];
+static const char *motor_names[NUMBER_OF_JOINTS] = {
+  "front left leg shoulder abduction motor",  "front left leg shoulder rotation motor",  "front left leg elbow motor",
+  "front right leg shoulder abduction motor", "front right leg shoulder rotation motor", "front right leg elbow motor",
+  "rear left leg shoulder abduction motor",   "rear left leg shoulder rotation motor",   "rear left leg elbow motor",
+  "rear right leg shoulder abduction motor",  "rear right leg shoulder rotation motor",  "rear right leg elbow motor"};
```
Comment (omichel) [inline projects/robots/boston_dynamics/spot/controllers/spot_moving_demo/spot_moving_demo.c:41]: I am wondering whether we should remove the "leg" keyword from the names here as it seems to be unnecessary to combine with shoulder or elbow, etc.

File: projects/robots/boston_dynamics/spot/controllers/spot_moving_demo/spot_moving_demo.c [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+static WbDeviceTag cameras[NUMBER_OF_CAMERAS];
+static const char *camera_names[NUMBER_OF_CAMERAS] = {"left head camera", "right head camera", "left flank camera",
+                                                      "right flank camera", "rear camera"};
+static WbDeviceTag leds[NUMBER_OF_LEDS];
+static const char *led_names[NUMBER_OF_LEDS] = {"left top led",          "left middle up led", "left middle down led",
+                                                "left bottom led",       "right top led",      "right middle up led",
+                                                "right middle down led", "right bottom led"};
+
+void step() {
```
Comment (omichel) [inline projects/robots/boston_dynamics/spot/controllers/spot_moving_demo/spot_moving_demo.c:50]:
```suggestion
static void step() {
```

File: docs/guide/spot.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+- `extension`: Extends the robot with new nodes.
+
+### Samples
+
+You will find the following sample in the folder: "[$WEBOTS\_HOME/projects/robots/boston\_dynamics/spot/worlds](https://github.com/cyberbotics/webots/tree/master/projects/robots/boston_dynamics/spot/worlds)"
+
+### [spot.wbt](https://github.com/cyberbotics/webots/tree/master/projects/robots/boston_dynamics/spot/worlds/spot.wbt)
+
+![spot.wbt.png](images/robots/spot/spot.wbt.thumbnail.jpg) This simulation shows a Spot robot in a simple environment.
+The robot is moving its right arm.
```
[the remark this answers is not in the document]
Comment (omichel) [reply docs/guide/spot.md:47]: I would say right leg...

[the remark this answers is not in the document]
Comment (DavidMansolino) [reply docs/guide/spot.md:47]: Indeed, I didn't updated this part after simplifying the controller.

**Comment by DavidMansolino** [comment]
> Shall we create new PBR appearances for the plastic padded and plastic rough instead of keeping them inside the spot proto folder?

Actually, for now, my rule for creating new appearance is to create one as soon as the texture is used in more than one world/proto, but if you think this textures are nice to showcase we can indeed create new appearances.

**Comment by omichel** [comment]
Yes, I believe plastic are very common appearances that could be useful in many new robots, so I would at least add the plastic rough one.

**Comment by omichel** [comment]
👀

**Pull request merged**